packages/backend-output-storage/src/stack_metadata_output_storage_strategy.ts

```diff
@@ -49,13 +50,32 @@ export class StackMetadataBackendOutputStorageStrategy
     backendOutputEntry: BackendOutputEntry
   ): void => {
     const version = backendOutputEntry.version;
+    let listsMap = this.lazyListValueMap.get(keyName);
 
-    Object.entries(backendOutputEntry.payload).forEach(([listName, value]) => {
-      let listsMap = this.lazyLists.get(keyName);
+    const metadata = this.stack.templateOptions.metadata
+      ? this.stack.templateOptions.metadata
+      : {};
+    const existingMetadataEntry = metadata[keyName];
+
+    if (existingMetadataEntry) {
+      if (existingMetadataEntry.version !== version) {
+        throw new Error(
+          `Metadata entry for ${keyName} at version ${existingMetadataEntry.version} already exists. Cannot add another entry for the same key at version ${version}.`
+        );
+      }
+    } else {
+      this.stack.addMetadata(keyName, {
+        version,
+        stackOutputs: Lazy.list({
+          produce: () => Array.from(listsMap ? listsMap.keys() : []),
+        }),
+      });
+    }
 
+    Object.entries(backendOutputEntry.payload).forEach(([listName, value]) => {
       if (!listsMap) {
         listsMap = new Map();
-        this.lazyLists.set(keyName, listsMap);
+        this.lazyListValueMap.set(keyName, listsMap);
```

Comment:
This can go right after line 53

Reply:
I put it here because Typescript is unhappy and thinks listsMap might be undefined in the `Object.entries.forEach` loop